### ISSUE TITLE
Validating the need for normalizr in client side apps 

### ISSUE THREAD

**Issue by emilchacko** [paularmstrong/normalizr#103]
I had this question for a long time.Don't know where to ask this question.So I'm posting it here.Looking for interesting inputs from normalizr users.

The main reason why nomalizr is required is because,the api responses are not not in a normalised format .Related entities are nested within the root entity itself.My question is why should REST/JSON api's return data in denormalised format in the first place .Why not do something like this .

Suppose this is a REST api to fetch all ticket and include the ticket contacts along with it ,ticket contacts has many-many relationship.

```
GET /api/tickets ? include="contact";
{
  list:[123,234] , //ticketIds
  tickets:{
   123:{
   id:123,
   subject:"My issue";
   fromConact : 345;
   }
   ...
 },
contacts:{

 345:{
 id:345,
 contactName:"john"
 }

 ...

}
}

```
**Advantages:**

- The client need not do the normalisation process .

- Less data to download for the client .

- Can model order of objects as separate data .Currently since most of the apis expose data as Json Arrays,its not clear if there is an ordering maintained in the array or is it just randomly ordered data in an array .Also an added advantage is that we can support multiple ordering of data.

**Disadvantages :** 
   

- If the client is not going to store the data in a cache then ,its a bit difficult to fetch the contact details from the json. 

- Need to follow this structure even for simple apis,where only a single data is being send to client or even if there is no related entity to be included  .eg.
   GET /api/tickets/123 ? include="contact"; 

To summarise my question  isn't normalisation the job of server ,rather than the client ? So shouldn't we just send normalised data to client and have a tool like reselect compute derive data or denormalise as required .May be I'm wrong please correct me ,if you guys think so.

**Comment by paularmstrong** [comment]
> To summarise my question isn't normalisation the job of server ,rather than the client ?

In an ideal scenario, yes. However, a lot of APIs are not in our control when writing applications that use them.

For example, say you want to write a GitHub issue reader. Issues come packed with users (and as `assignee`) nested in them: https://developer.github.com/v3/issues/.

**Comment by emilchacko** [comment]
@paularmstrong : Thanks for the quick response .Your right ,its a valid point .But apart from that ,do you think of any case where the REST api should respond with denormalised data .Since this is the general trend I see in most API ,like github,facebook,twitter ect .Which makes me think ,is there something wrong with normalising the data on server ,or may be they haven't thought about it or adopted this way for the sake of simplicity.

**Comment by paularmstrong** [comment]
There's absolutely nothing wrong with normalizing your data before you send it to clients. If you are able to, I'd definitely encourage it! If you want to take things one step further, I'd recommend following the [JSON API Spec](http://jsonapi.org/).

**Comment by andrewmclagan** [comment]
@paularmstrong is normalizer needed if the responses follow the JSON API specifications?

**Comment by paularmstrong** [comment]
¯\_(ツ)_/¯ I've never actually used it.

**Comment by andrewmclagan** [comment]
Haha will report back my findings

**Comment by paularmstrong** [comment]
To clarify, by "it", I mean the JSON API spec. It looks great, but I've just never been blessed enough to work with an API with a strictly defined spec.

**Comment by andrewmclagan** [comment]
Yeah it looks great,  although a little data heavy to me
